virttest/utils_qemu.py

```diff
@@ -19,10 +32,104 @@ def get_qemu_version(bin_path):
-    output = process.system_output("%s -version" % bin_path,
-                                   verbose=False,
-                                   ignore_status=True).decode()
+    output = _get_qemu_info(bin_path, "-version")
     matches = QEMU_VERSION_RE.match(output)
     if matches is None:
         raise OSError('Unable to get the version of qemu')
     return matches.groups()
+
+
+def get_machines_mapping(bin_path):
+    """
+    Return all machine type supported by qemu
+
+    :param bin_path: Path to qemu binary
+    :return: A dict of all machine types supported by qemu
+    """
+    output = _get_qemu_info(bin_path, r"-machine \?")
+    machines = re.findall(r"^([0-9a-zA-Z._-]+).*[^:]$", output, re.M)
+    machines_mapping = dict(zip(machines, machines))
+    has_alias_machines = re.findall(r"^([0-9a-zA-Z._-]+).*\(alias of (\S+)\)",
+                                    output, re.M)
+    machines_mapping.update(dict(has_alias_machines))
+    return machines_mapping
+
+
+def get_supported_machines(bin_path):
+    """
+    Return all machine type supported by qemu
+
+    :param bin_path: Path to qemu binary
+    :return: A list of all machine types supported by qemu
+    """
+    return list(get_machines_mapping(bin_path).keys())
+
+
+def get_correct_machine_type(bin_path, machine_type):
+    """
+    Return the correct machine type
+
+    :param bin_path: Path to qemu binary
+    :param machine_type: qemu machine type
+    :return: Correct machine type
+    """
+    try:
+        return get_machines_mapping(bin_path)[machine_type]
+    except KeyError:
+        raise KeyError("qemu does not support this machine type: %s"
+                       % machine_type)
+
+
+def get_supported_devices(bin_path, category=None):
+    """
+    Return all devices supported by qemu
+
+    :param bin_path: Path to qemu binary
+    :param category: device category (e.g. 'USB', 'Network', 'CPU')
+    :return: A set of devices supported by qemu
+    """
+    output = _get_qemu_info(bin_path, r"-device \?")
+    qemu_devices = {}
+    for device_info in output.split("\n\n"):
+        device_type = re.match(r"([A-Z]\S+) devices:", device_info).group(1)
+        qemu_devices[device_type] = set(re.findall(r'(?:name|alias) "(\S+)"',
+                                                   device_info))
+    if category:
+        return qemu_devices[category]
+    return set.union(*qemu_devices.values())
+
+
+def get_maximum_cpus(bin_path, machine_type):
+    """
+    Return max CPUs supported by specified machine type
+
+    :param bin_path: Path to qemu binary
+    :param machine_type: One machine type supported by qemu
+    :raise ValueError: If unable to get that
+    :return: Maximum value of vCPU
+    """
+    invalid_maximum = 2 ** 31 - 1
+    correct_machine_type = get_correct_machine_type(bin_path, machine_type)
+    output = _get_qemu_info(bin_path, r"-machine %s -smp maxcpus=%d" %
+                            (correct_machine_type, invalid_maximum))
```

Comment:
Not sure if it is handled, the maximum cpus for q35 is 384, but it only works when add an IOMMU.

# /usr/libexec/qemu-kvm -smp 4,maxcpus=400 -M q35
qemu-kvm: Invalid SMP CPUs 400. The max CPUs supported by machine 'pc-q35-rhel7.6.0' is 384

# /usr/libexec/qemu-kvm -smp 4,maxcpus=384 -M q35
VNC server running on ::1:5900
qemu-kvm: current -smp configuration requires Extended Interrupt Mode enabled. You can add an IOMMU using: -device intel-iommu,intremap=on,eim=on

Reply:
```
invalid_maximum = 2 ** 31 - 1
# /usr/libexec/qemu-kvm -smp 4,maxcpus=2147483647 -M q35
qemu-kvm: Invalid SMP CPUs 2147483647. The max CPUs supported by machine 'pc-q35-rhel8.1.0' is 384
```

Reply:
We need to handle it ourselves